src/compiler/checker.ts

```diff
@@ -18205,16 +18409,110 @@ namespace ts {
+                    if (declaration.kind === SyntaxKind.ClassDeclaration || declaration.kind === SyntaxKind.InterfaceDeclaration) {
+                        const typeParameterNodes = (<ClassDeclaration | InterfaceDeclaration>declaration).typeParameters;
+                        const numTypeParameters = typeParameterNodes ? typeParameterNodes.length : 0;
+                        if (maxTypeArgumentCount === -1) {
+                            // For the first declaration, establish the initial maximum and
+                            // minimum type argument counts. These only change when we
+                            // encounter default type arguments.
+                            maxTypeArgumentCount = numTypeParameters;
+                            minTypeArgumentCount = numTypeParameters;
+                        }
+
+                        if (typeParameterNodes) {
+                            if (!typeParameters) {
+                                typeParameters = [];
+                            }
+
+                            for (let i = 0; i < typeParameterNodes.length; i++) {
+                                if (typeParameterNodes[i].default) {
+                                    // When we encounter a type parameter with a default, establish
+                                    // new minimum and maximum type arguments if necessary.
+                                    if (minTypeArgumentCount > i) {
+                                        minTypeArgumentCount = i;
+                                    }
+                                    if (maxTypeArgumentCount < i + 1) {
+                                        maxTypeArgumentCount = i + 1;
+                                    }
+                                }
+                                if (typeParameters.length <= i) {
+                                    // When we encounter a new type parameter at this position,
+                                    // get the declared type for the type parameter. If another
+                                    // declaration attempts to establish a type parameter with a
+                                    // different name or constraint than the first one we find,
+                                    // we will report an error when checking the type parameters.
+                                    typeParameters[i] = getDeclaredTypeOfTypeParameter(getSymbolOfNode(typeParameterNodes[i]));
+                                }
+                            }
+                        }
                     }
                 }
+                if (maxTypeArgumentCount === -1) {
+                    maxTypeArgumentCount = 0;
+                }
+                if (minTypeArgumentCount === -1) {
+                    minTypeArgumentCount = maxTypeArgumentCount;
+                }
+                if (typeParameters && typeParameters.length > maxTypeArgumentCount) {
+                    // Trim the type parameters to the maximum length

```

Comment:
I don't think it works to trim like this. You might have members that reference the trimmed type parameter and those members would never be properly instantiated. In other words, the trimmed type parameters could "leak".